plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,539 @@
+                .whenComplete(
+                    response -> log.info("node id:{} insert retry times successfully", localNodeId),
+                    saveLatestDeployRetryTimesStep::onFailure
+                );
+        }, getDeployRetryTimesStep::onFailure);
+    }
+
+    /**
+     * auto undeploy all the models under the node id <br>
+     *
+     * @param localNodeId local node id
+     * @param modelIdAndNodeIds model ids
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void unDeployModel(String localNodeId, Map<String, Set<String>> modelIdAndNodeIds) throws MLException {
+        StepListener<MLUndeployModelNodesResponse> unDeployModelStep = new StepListener<>();
+
+        ActionListener<MLUndeployModelNodesResponse> listener = ActionListener
+            .wrap(
+                response -> log.info("the models is auto undeploying under the node {} ", localNodeId),
+                exception -> log.error("fail to undeploy models under the node id: {} \nthe reason is: {}", localNodeId, exception)
+            );
+        String[] modelIds = modelIdAndNodeIds.keySet().toArray(new String[] {});
+
+        autoUnDeployModelByNodeAndModelId(localNodeId, modelIds, listener);
+
+        unDeployModelStep.whenComplete(unDeployModelStep::onResponse, unDeployModelStep::onFailure);
+    }
+
+    /**
+     * auto deploy 1 model under the local node id <br>
+     *
+     * @param nodeIds node ids
+     * @param modelId model id
+     * @param listener listener
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoDeployModelByNodeAndModelId(String[] nodeIds, String modelId, ActionListener<MLDeployModelResponse> listener)
+        throws MLException {
+        MLDeployModelRequest deployModelRequest = new MLDeployModelRequest(modelId, nodeIds, false, false);
+
+        client.execute(MLDeployModelAction.INSTANCE, deployModelRequest, listener);
+    }
+
+    /**
+     * auto undeploy 1 model under the local node id <br>
+     *
+     * @param localNodeId node id
+     * @param modelIds all the model id
+     * @param listener listener
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoUnDeployModelByNodeAndModelId(String localNodeId, String[] modelIds, ActionListener<MLUndeployModelNodesResponse> listener)
+        throws MLException {
+        MLUndeployModelNodesRequest undeployModelNodesRequest = new MLUndeployModelNodesRequest(new String[] { localNodeId }, modelIds);
+
+        client.execute(MLUndeployModelAction.INSTANCE, undeployModelNodesRequest, listener);
+    }
+
+    /**
+     * get all models from the index <br>
+     *
+     * @param searchResponseActionListener listener for search response
+     */
+    @VisibleForTesting
+    void queryModels(ActionListener<SearchResponse> searchResponseActionListener) {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+
+        QueryBuilder queryBuilder = QueryBuilders
```

Comment:
Is it possible to use only a must condition, e.g. LOADED/PARTIAL_LOADED to build the query? This needs less code, right?

Reply:
I found the `termQuery` or `termsQuery` don't work in unit test.So I changed it to `marchPhraseQuery`, let one by one filter